File: medium/498.java
// import java.util.*;

// class Solution {
//     public int[] findDiagonalOrder(int[][] mat) {

//         HashMap<Integer, List<Integer>> map = new HashMap<>();

//         for (int i = 0; i < mat.length; i++)
//             for (int j = 0; j < mat[0].length; j++)
//                 map.computeIfAbsent(i + j, k -> new ArrayList<>()).add(mat[i][j]);

//         List<Integer> resultList = new ArrayList<>();
//         boolean flip = true;
//         for (int k = 0; k <= mat.length + mat[0].length - 2; k++) {
//             List<Integer> diagonal = map.get(k);
//             if (diagonal == null)
//                 continue;
//             if (flip)
//                 Collections.reverse(diagonal);
//             resultList.addAll(diagonal);
//             flip = !flip;
//         }
//         int[] result = new int[resultList.size()];
//         for (int i = 0; i < resultList.size(); i++)
//             result[i] = resultList.get(i);

//         return result;
//     }
// }

// // 0 1 2
// // 3 4 5
// // 6 7 8

// // 0 -> 0
// // 1 -> 1 3
// // 2 -> 2 4 6
// // 3 -> 5 7
// // 4 -> 8

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        if (mat == null || mat.length == 0)
            return new int[0];

        int m = mat.length, n = mat[0].length;
        int[] result = new int[m * n];

        int row = 0, col = 0, dir = 1;
        int idx = 0;

        while (idx < m * n) {
            result[idx++] = mat[row][col];

            if (dir == 1) { // moving up-right
                if (col == n - 1) { // hit right boundary
                    row++;
                    dir = -1;
                } else if (row == 0) { // hit top boundary
                    col++;
                    dir = -1;
                } else {
                    row--;
                    col++;
                }
            } else { // down
                if (row == m - 1) {
                    col++;
                    dir = 1;
                } else if (col == 0) {
                   row++;
                   dir = 1;
                }
                else{
                    row++;
                    col--;
                }
            }
        }

        return result;
    }
}

File: medium/498.py
# from collections import defaultdict


# class Solution:
#     def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
#         diag = defaultdict(list)

#         for i in range(len(mat)):
#             for j in range(len(mat[0])):
#                 diag[i+j].append(mat[i][j])

#         result = []
#         flip = True
#         for i in range(len(mat) + len(mat[0]) - 1):
#             diagonal = diag.get(i, [])
#             if flip:
#                 diagonal.reverse()

#             result.extend(diagonal)
#             flip = not flip

#         return result


class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m = len(mat)
        n = len(mat[0])

        row = 0
        col = 0
        idx = 0
        result = [0] * (m * n)
        dir = 1
        while idx < m * n:
            result[idx] = mat[row][col]
            idx += 1
            if dir == 1:
                if col == n-1:
                    row += 1
                    dir = -1
                elif row == 0:
                    col += 1
                    dir = -1
                else:
                    row -= 1
                    col += 1

            elif dir == -1:
                if row == m-1:
                    col += 1
                    dir = 1

                elif col == 0:
                    row += 1
                    dir = 1

                else:
                    row += 1
                    col -= 1

        return result
